Mathlib/Util/Term/Basic.lean

```diff
@@ -0,0 +1,47 @@
+/-
+Copyright (c) 2021 Gabriel Ebner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gabriel Ebner, Yuyang Zhao
+-/
+import Lean.Elab.SyntheticMVars
+import Lean.Meta.Tactic.Delta
+import Mathlib.Lean.Expr.Basic
+import Mathlib.Util.Term.Beta
+
+/-!
+# Basic term elaborators
+-/
+
+namespace Mathlib.Util.Term.Basic
+
+open Lean Elab Term Meta
+
+/-- `delta% t` elaborates to a head-delta reduced version of `t`. -/
+elab "delta% " t:term : term <= expectedType => do
+  let t ← withSynthesize do
+    elabTermEnsuringType t expectedType
+  synthesizeSyntheticMVars
```

Comment:
Any reason you changed the original implementation? I see two changes:
1. `elabTerm` -> `elabTermEnsuringType`
2. this is wrapped in `withSynthesize`.

I don't think using `elabTermEnsuringType` is justified, for any of these term elaborators.

`withSynthesize` is arguably useful, and it does look like you have a test case exercising it. Still, it's rather strong, and it might be better configuring it with `PostponeBehavior.partial`.